src/agent/src/device.rs

```diff
@@ -56,9 +56,50 @@ struct DevIndexEntry {
 #[derive(Debug)]
 struct DevIndex(HashMap<String, DevIndexEntry>);
 
+fn filename_is_pci_addr(f: &std::ffi::OsStr) -> bool {
+    lazy_static! {
+        static ref PCIADDR: Regex =
+            Regex::new(r"^[0-9a-f]{4}:[0-9a-f]{2}:[0-9a-f][.][0-7]$").unwrap();
+    }
+
+    let s = match f.to_str() {
+        Some(n) => n,
+        None => return false,
+    };
+    PCIADDR.is_match(s)
+}
+
 #[instrument]
-pub fn rescan_pci_bus() -> Result<()> {
-    online_device(SYSFS_PCI_BUS_RESCAN_FILE)
+pub fn rescan_pci_shpc() -> Result<()> {
+    // With certain guest types, SHPC is the only available PCI
+    // hotplug option.  Unfortunately, SHPC has a mandated 5s delay
+    // (designed for an actual person physically plugging a card).
+
+    // To work around that, we force a PCI rescan when SHPC is in use,
+    // which will detect the device earlier.  This rescan can cause a
+    // number of other problems, so we want to avoid it if we're not
+    // using SHPC
+
+    let dir = match fs::read_dir(SYSFS_PCI_SHPCHP_DRIVER) {
```

Comment:
How about making this function accept a file parameter: `rescan_pci_shpc(file: &str) -> Result<()>` such that we would call it as `rescan_pci_shpc(SYSFS_PCI_SHPCHP_DRIVER)`. That way, you could write some unit tests for it.

Reply:
Good call, I've added a test for the next spin.